### PR TITLE
Add href support to resource_option

### LINKED ISSUE #315: For sync there's no remote href option nor help text about what's expected for the remote option
What does pulp-cli expect for `--remote` and where is the option for href?

```
$ pulp file repository sync --help
Usage: pulp file repository sync [OPTIONS]

Options:
  --name TEXT    Name of the repository
  --href TEXT    HREF of the repository
  --remote TEXT
  --help         Show this message and exit.
```

### PR DESCRIPTION
**Pull request by mdellweg**
fixes #315

### CODE REVIEW

**Comment by mdellweg** [comment]
This is currently only implemented in "pulp file repository *" commands to showcase.

File: pulpcore/cli/common/generic.py [hunk 1]
```diff
@@ -256,10 +258,35 @@ def _option_callback(
...
             return value
 
-        split_value = value.split(":", maxsplit=2)
-        while len(split_value) < 3:
-            split_value.insert(0, "")
-        plugin, resource_type, identifier = split_value
+        pulp_href: Optional[str] = None
+        entity: Optional[EntityDefinition] = None
+
+        if value.startswith("/"):
```
Comment (daviddavis) [inline pulpcore/cli/common/generic.py:264]: I wonder if this shouldn't check for href_pattern instead and then just assume it's a name if it doesn't match:

```
$ pulp file remote create --name /test/ --url https://fixtures.pulpproject.org/file/
$ pulp file repository create --name /test/ --remote /test/
Error: '/test/' is not a valid href for remote.
```

Comment (mdellweg) [reply pulpcore/cli/common/generic.py:264]: That wouldn't solve the problem where the name actually looks like a valid href.
But you can always use: `$ pulp file repository create --name /test/ --remote ::/test/`

**Review by daviddavis — APPROVED** [review]

**Comment by mdellweg** [comment]
@daviddavis I extended the use of this feature to all plugins sync and export{,er}s. You might want to revisit.

**Review by daviddavis — APPROVED** [review]

**Comment by daviddavis** [comment]
I had one thought (not necessarily for this PR): what if we stored the href_pattern on the context instead of having to define it everywhere?

**Comment by mdellweg** [comment]
> I had one thought (not necessarily for this PR): what if we stored the href_pattern on the context instead of having to define it everywhere?

I'm happy to move it around (to generic?). But for the more generic options, what should be _the_ Context to store it on?

**Comment by daviddavis** [comment]
Maybe for generic options, the href_pattern must be supplied when calling resource_option?

**Comment by mdellweg** [comment]
I moved it into the base context. It seemd to match best there.

File: pulpcore/cli/core/export.py [hunk 2]
```diff
@@ -16,16 +18,34 @@
 from pulpcore.cli.core.context import PulpExportContext, PulpExporterContext
 
 _ = gettext.gettext
+REPO_HREF_PATTERN = r"^/pulp/api/v3/repositories/(?P<plugin>\w+)/(?P<resource_type>\w+)/"
```
Comment (daviddavis) [inline pulpcore/cli/core/export.py:21]: Why define this hear instead of using `PulpRepositoryContext.HREF_PATTERN`?

**Comment by daviddavis** [comment]
I was thinking that `_option_callback` in `resource_option` would have some logic like:

```python
if href_pattern is None and hasattr(ctx, HREF_PATTERN):
    href_pattern = ctx.HREF_PATTERN
```

File: pulpcore/cli/core/export.py (hunk 2 above)
Comment (mdellweg) [reply pulpcore/cli/core/export.py:21]: forgot about that.

**Comment by mdellweg** [comment]
> I was thinking that `_option_callback` in `resource_option` would have some logic like:
> 
> ```python
> if href_pattern is None and hasattr(ctx, HREF_PATTERN):
>     href_pattern = ctx.HREF_PATTERN
> ```

The problem is, it needs the pattern before determining the Context class to instanciate.

**Comment by daviddavis** [comment]
Ah, I see. That kind of defeats part of the purpose of my suggestion to have it on context then.

**Comment by mdellweg** [comment]
> Ah, I see. That kind of defeats part of the purpose of my suggestion to have it on context then.

It still feels like the right place.

**Review by daviddavis — APPROVED** [review]

**Pull request merged**